pudb/debugger.py

```diff
@@ -878,17 +881,46 @@ def change_rhs_box(name, index, direction, w, size, key):
 
         # {{{ variables listeners
 
+        def get_inspect_info(id_path, read_only=False):
+            return (self.get_frame_var_info(read_only)
+                    .get_inspect_info(id_path, read_only))
+
+        def collapse_current(var, pos, iinfo):
+            if iinfo.show_detail:
+                # Simple case: collapse current variable
+                iinfo.show_detail = False
+            else:
+                # Complex case: collapse parent/container variable
+                desired_prefix = var.prefix[:-len(ValueWalker.PREFIX)]
+                try:
+                    # Walk backwards to find first shorter prefix
+                    parent = next(
+                        variable_widget
+                        for variable_widget in reversed(self.locals[:pos+1])
+                        if variable_widget.prefix == desired_prefix
+                    )
+                except StopIteration:
+                    # No parent found, so don't do anything
+                    pass
+                else:
+                    p_iinfo = get_inspect_info(parent.id_path)
+                    p_iinfo.show_detail = False
+                    self.var_focus_index = self.locals.index(parent)
```

Comment:
I agree that this works currently, but I don't like that this uses a presentation-related property (`prefix`) for logic purposes (as there's no guarantee that this string will always look the way it does now). Could you refactor this/the `VariableWidget` so that the `VariableWidget` carries a (structural) `nesting_level` attribute instead (and, ideally, compute the prefix from that)?

Reply:
Yup I'm happy to do so. I initially avoided this approach to try to minimize the impact of my patch but I agree this would be better!